src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,520 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "jitpch.h"
+#include "rangecheckcloning.h"
+
+// This file contains the definition of the "Range check cloning" phase.
+//
+// The goal of this phase is to pick up range checks which were not optimized by the
+// range check optimization phase and clone them to have "fast" and "slow" paths.
+// This is similar to what the "Loop Cloning" phase does for loops. Example:
+//
+//   arr[i + 1] = x;
+//   arr[i + 3] = y;
+//   arr[i + 5] = z;
+//   arr[i + 8] = w;
+//
+// assertprop/rangecheck phases give up on the above bounds checks because of the
+// increasing offsets and there are no assertions that they can rely on.
+// This phase handles such cases by cloning the entire block (only the affected statements
+// to be precise) into "fast" and "slow" under a "cloned" condition:
+//
+//   if (i >= 0 && i < arr.Length - 8)
+//   {
+//       // Fast path
+//       arr[i + 1] = x; // no bounds check
+//       arr[i + 3] = y; // no bounds check
+//       arr[i + 5] = z; // no bounds check
+//       arr[i + 8] = w; // no bounds check
+//   }
+//   else
+//   {
+//       // Slow path
+//       arr[i + 1] = x; // bounds check
+//       arr[i + 3] = y; // bounds check
+//       arr[i + 5] = w; // bounds check
+//       arr[i + 8] = w; // bounds check
+//   }
+//
+// The phase scans all statements in a block and groups the bounds checks based on
+// "Base Index and Length" pairs (VNs). Then the phase takes the largest group and
+// clones the block to have fast and slow paths.
+
+//------------------------------------------------------------------------------------
+// Initialize: Initialize the BoundsCheckInfo with the given bounds check node.
+//    and perform some basic legality checks.
+//
+// Arguments:
+//    comp             - The compiler instance
+//    statement        - The statement containing the bounds check
+//    bndChkNode       - The bounds check node
+//    bndChkParentNode - The parent node of the bounds check node (either null or COMMA)
+//
+// Return Value:
+//    true if the initialization was successful, false otherwise.
+//
+bool BoundsCheckInfo::Initialize(const Compiler*   comp,
+                                 Statement*        statement,
+                                 GenTreeBoundsChk* bndChkNode,
+                                 GenTree*          bndChkParentNode)
+{
+    idxVN = comp->vnStore->VNConservativeNormalValue(bndChkNode->GetIndex()->gtVNPair);
+    lenVN = comp->vnStore->VNConservativeNormalValue(bndChkNode->GetArrayLength()->gtVNPair);
```

Comment:
I think since the length is removed we would want to include the exception set here, for the length only.

Reply:
Ended up introducing my own optRemoveRangeCheck without relaxations for side-effects like we've discussed offline.